flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java

```diff
@@ -137,7 +137,13 @@ public static long getSizeOfFreeHeapMemoryWithDefrag() {
 	 */
 	public static long getSizeOfFreeHeapMemory() {
 		Runtime r = Runtime.getRuntime();
-		return r.maxMemory() - r.totalMemory() + r.freeMemory();
+		long maxMemory = r.maxMemory();
+		if (maxMemory == Long.MAX_VALUE) {
+			// workaround for some JVM versions
+			return r.freeMemory();
```

Comment:
We could fall back to a conservative value and print a warning that the memory has to be set manually. To set the memory manually, the user has to go through the LocalExecutor which is not a nice out of the box experience.